library/base.go

```diff
@@ -0,0 +1,227 @@
+
+func (a *AuthZRules) getControlGroup(name string) (*ControlGroup, error) {
+	for _, cg := range a.ControlGroups {
+		if cg.Name == name {
+			return &cg, nil
+		}
+	}
+	return nil, errors.New("no such control group name")
+}
+
+func (r *RunnerImpl) internalCleanup(puuid Process) {
+	log.Println("looking up process", puuid)
+	fn := fmt.Sprintf("/sys/fs/cgroup/%s", puuid)
+	if err := os.RemoveAll(fn); err != nil {
+		log.Fatalln("cannot clean up cgroup dir", fn, err)
+	}
+
+	// clean up temp dir
+	if rp, ok := r.getRunningProcess(puuid); ok {
+		log.Println("cleaning up", rp.tempdir)
+		if err := os.RemoveAll(rp.tempdir); err != nil {
+			log.Fatalln("cannot clean up temp dir", rp.tempdir, err)
+		}
+	} else {
+		log.Println("could not find process", puuid)
+	}
+}
+
+func cleanup(r *RunnerImpl, puuid Process) func() {
+	return func() {
+		r.internalCleanup(puuid)
+	}
+}
+
+func (r *RunnerImpl) addRunningProcess(puuid Process, rp *runningProcess) {
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+	r.running[puuid] = rp
+}
+
+func (r *RunnerImpl) getRunningProcess(puuid Process) (*runningProcess, bool) {
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+	rp, ok := r.running[puuid]
+	return rp, ok
+}
+
+func (r *RunnerImpl) removeRunningProcess(puuid Process) {
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+	delete(r.running, puuid)
+}
+
+func (r *RunnerImpl) Run(ctx context.Context, client string, cmd string, args ...string) (Process, error) {
+	cgn, ok := r.authz.ClientToCGroup[client]
+	if !ok {
+		return "", errors.New("authorization failed - no cgroup defined for this client")
+	}
+	cg, err := r.authz.getControlGroup(cgn)
+	if err != nil {
+		return "", err
+	}
+
+	puuid := Process(uuid.New().String())
+
+	dir, err := ioutil.TempDir("", string(puuid)+"_*")
+	if err != nil {
+		return "", err
+	}
+
+	os.Chdir(dir)
+
+	if err := r.cgi.MakeCgroup(cg, puuid); err != nil {
+		log.Fatalln("unable to make cgroup:", err)
+	}
+
+	cgDir := fmt.Sprintf("/sys/fs/cgroup/%s", puuid)
+
+	var buf SafeBuffer
+	ps, st, done, err := r.cgi.Run(ctx, r.wrapperPath, cg.SysProcAttr, &buf, cleanup(r, puuid), cgDir, cmd, args...)
+	if err != nil {
+		r.internalCleanup(puuid)
+		return "", err
+	}
+
+	if ps.Process != nil {
+		r.addRunningProcess(puuid, &runningProcess{cmd: ps, output: &buf, tracker: st, pid: ps.Process.Pid, tempdir: dir, done: done})
+		return puuid, nil
+	}
+	r.internalCleanup(puuid)
+	return "", errors.New("process did not launch")
+}
+
+func (r *RunnerImpl) safeGetRunningProcess(p Process) (*runningProcess, error) {
+	rp, ok := r.running[p]
+	if !ok {
+		return nil, errors.New("process not found")
+	}
+	return rp, nil
+}
+
+func (r *RunnerImpl) GetStatus(ctx context.Context, p Process) (Status, int, error) {
+	rp, err := r.safeGetRunningProcess(p)
+	if err != nil {
+		return Unavailable, -1, err
+	}
+	st, ec := rp.tracker.GetStatus()
+	return st, ec, nil
+}
+
+func (r *RunnerImpl) StreamOutput(ctx context.Context, p Process, sender Sender) error {
```

Comment:
Internally, you've done a nice job at using the standard `io.Reader` and `io.Writer` interfaces. This is great - your buffer implementation can now integrate with a lot of  existing Go code with minimal effort.

When it comes to the public API your library presents to streamers, however, you expose a custom `Sender` interface instead of sticking to reusable and standard `io` interfaces. I wonder how difficult it would be to do away with the `Sender` interface here.

Reply:
Yeah, I wasn't happy with the `Sender` thing myself. It would be best to use the `safeBuffer` again (that same 1:N fanout) but `StreamOutput()` doesn't return until the transfer is finished so I can't return a `os.Reader` to the caller. 

Any suggestions for passing in a `Reader`?